### DISCUSSION TITLE
RJE via e-mail

### DISCUSSION THREAD

**Discussion by fbelnet** [kej715/DtCyber#161] [General]
Hello from an old controlfreak.org member.

I would like to implement a RJE process polling any incoming e-mail sent to rje@cdc.host.
Then passing its text body as an input job and then mailing back the output to the e-mail sender.

Fred G.

**Comment by kej715** [comment]
Hi Fred,

Looks like an interesting project. The rje-station subdirectory of the
DtCyber repo contains Javascript, HTML, and Node.js code that emulates BSC,
HASP, and MODE4 protocols, and it has a command line RJE client as well as
a web-based RJE interface. You should be able to use the BSC, HASP, and
MODE4 modules to implement an email interface, perhaps deriving from the
command line RJE client. Especially if you're familiar with Node.js
programming, it should be straightforward to find Node.js modules for
reading/sending email, and then you can use the RJE modules in the
rje-station subdirectory to send email bodies as card decks and construct
email bodies from line printer and punch output.

cheers,
Kevin

**Reply by fbelnet** [reply]
Hey Kevin,

Unfortunately I know more about programming in compass or sympl than Node.js 😊 

Hopefully someone will have some time to look at this and maybe find a way.

J. Phillip Draughon (controlfreak) had an implementation of this feature for NOS 1.4 but I have looked at my archives and cannot find anything.

**Comment by philclaridge** [comment]
Hi Fred,
This is something I can help with later in the year. 

The last time I did RJE was via a real card reader in the 1970s so I’ll need some advice on how to get an emulation setup and working with a test remote job that runs under the emulation/OS version you want. 

Note I run the emulations interactively so don’t need any help with compiling DtCyber and getting a basic machine running.

Phil

**Comment by kej715** [comment]
Hi Phil,

The DtCyber repo has subdirectory rje-station. Among other things, this
provides a web service that interfaces with RBF on NOS 2 and EI200 on
KRONOS and NOS 1. Furthermore, the ready-to-run packages for these
operating systems launch this web service automatically at DtCyber startup
time. I recommend leveraging this web service to build an RJE-by-mail
application. The web service greatly facilitates RJE interactions and
abstracts the large differences between HASP (i.e., RBF on NOS2) and MODE4
(i.e., EI200 on KRONOS and NOS 1). Also, DtCyber's direct TCP
interfaces for HASP and MODE4 are built atop BSC emulation (i.e, BSC over
TCP). This BSC emulation is compatible with the implentation provided by
the Hercules IBM mainframe emulator, and it is quite timing-sensitive. It
works well over a LAN, but probably won't work as reliably over WAN (i.e.,
public internet) connections. This is another issue resolved by the web
service. The web service runs on the same machine as DtCyber and handles
the low level HASP/BSC and MODE4/BSC. It presents a typical stateless (and
timing-insensitive) interface to its clients.

The web service interface currently is not documented. I'll fix that.

cheers,
Kevin

**Comment by kej715** [comment]
The RJE web service API is now documented. See:


https://github.com/kej715/DtCyber/blob/main/rje-station/README.md#using-rjews
